Add lowering for updated _scaled_mm (fixing submodules)

yangsiyu007 opened the pull request:
Add the Inductor lowering for `torch._scaled_mm`, whose API was last updated in https://github.com/pytorch/pytorch/pull/128683. 

The lowering does:
- for tensor-wise scaling, auto-tune between the default ATen kernel (cuBLAS) and Triton kernel configurations.
- for row-wise scaling, auto-tune between the default ATen kernel (CUTLASS kernel added in https://github.com/pytorch/pytorch/pull/125204) and Triton kernel configurations.

The Triton kernel template is based on https://github.com/htyu/FBGEMM/commit/3ad9031d0281550cd5e7a1e953e109f020f9bd8a (D56337896) by @choutim, without using SPLIT_K, and that of mm `torch/_inductor/kernel/mm.py`

## Testing:
- Logging shows max-autotune tuning (`AUTOTUNE scaled_mm`) for both tensor-wise and row-wise scaling when called with the two scaling types.
- Row-wise scaling allows operator fusion between preceding pointwise/reduction op and amax/cast:
    - output code Evaluating m=256, n=256, k=256, fusion_case='pointwise', scaling_mode='row'
        - P1477224245 - 2 kernels
    - output code Evaluating m=2048, n=256, k=2048, fusion_case='reduction', scaling_mode='row'
        - P1477227340 - 2 kernels

- UT `python test/inductor/test_fp8.py -- TestFP8Lowering`

## Benchmarking

Eager/compiled tensor-wise/row-wise scaling for various shapes: 
https://docs.google.com/spreadsheets/d/1VfWEVuyrwoWysfbS0_u2VHJ-PsdWkF1qIsiD60AzTes/edit?gid=2113587669#gid=2113587669
- Some of the “compiled” cases are slightly slower than “eager”. It’s because max-autotune selected the ATen kernel in the compiled case, and I think the discrepancy is variance.

Eager/compiled tensor-wise/row-wise scaling with pointwise/reduction preceding op for various shapes: 
https://docs.google.com/spreadsheets/d/1Nv07NrdffQIoDeMjo9E0V-E-EYrEN0WysO_bn1bc6ns/edit?gid=1715488446#gid=1715488446


## Questions for reviewers:
- Should the type of the accumulator `ACC_TYPE` always be in float32? If not, where is this type set (output layout?)?

## Todo:
- Make the Triton template use the improved persistent kernel version (https://github.com/pytorch/FBGEMM/pull/2735 by @htyu) 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

yangsiyu007 committed:
Add lowering for updated _scaled_mm.

yangsiyu007 committed:
Run linter.

yangsiyu007 committed:
Import triton only for type checking

yangsiyu007 committed:
Suppress linter error because cannot import triton

vkuzo commented:
> Should use_fast_accum be default False (as it is in aten/src/ATen/native/native_functions.yaml)? Probably should be False

Matching the setting from the user via `torch._scaled_mm` would be good.

vkuzo commented on test/inductor/test_fp8.py:
```diff
@@ -346,6 +368,67 @@ def ln_fp8(x: Tensor, scale: Tensor, amax_buffer: Tensor):
 
+@instantiate_parametrized_tests
+class TestFP8Lowering(TestCase):
+    @unittest.skipIf(TEST_WITH_ROCM, "FP8 is not supported on ROCM")
```
cc @drisspg , thoughts on how far away we are from also supporting the nuz float8 flavors here?

vkuzo commented on test/inductor/test_fp8.py:
```diff
@@ -346,6 +368,67 @@ def ln_fp8(x: Tensor, scale: Tensor, amax_buffer: Tensor):
+    def test_tensorwise_scaling(self, dtype: torch.dtype, shape: str, has_bias: bool):
+        device = "cuda"
+        dtype_float8 = torch.float8_e4m3fn
+        use_fast_accum = True
```
should we test both values?

vkuzo commented on test/inductor/test_fp8.py:
```diff
@@ -346,6 +368,67 @@ def ln_fp8(x: Tensor, scale: Tensor, amax_buffer: Tensor):
+            w_inverse_scale,
+            bias,
+        )
+        torch.testing.assert_close(y_eager, y_compiled, rtol=5e-1, atol=5e-1)
```
does this match how other matmul codegen cases in inductor are tested?

yangsiyu007 replied:
`rtol=5e-1, atol=5e-1` is the most strict used elsewhere in this file e.g. `test_valid_cast()`. In the Triton matmul [tutorial](https://triton-lang.org/main/getting-started/tutorials/03-matrix-multiplication.html#sphx-glr-getting-started-tutorials-03-matrix-multiplication-py), for fp8 it tested `atol=0.125, rtol=0` for M, N, K = 512 as a reference. Let me know if you have concerns!

vkuzo replied:
the float8 test in the triton tutorial is comparing PyTorch bf16 matmul to triton float8 matmul.  The comparison here is between eager mode float8 matmul and compiled float8 matmul - should that be tested with a higher tolerance?

yangsiyu007 replied:
You're right, let me check the actual eager vs compiled difference for some shapes to see what's a good strict tolerance.

yangsiyu007 replied:
@vkuzo updated the tolerance for each after checking the actual differences for the shapes tested. I initially read `5e-1` as `1e-5` in my head which was not helpful...

ipiszy commented on torch/_inductor/kernel/mm_common.py:
```diff
@@ -179,9 +179,109 @@ def filtered_configs(
 )
 
-# Create filtered list of configs based on cond evaluation
+scaled_mm_kernel_configs = [
```
nit: add a comment to describe the meaning of each field.

yangsiyu007 replied:
There is a comment describing the fields above `mm_kernel_configs` on line 119.

ipiszy commented on torch/_inductor/kernel/mm_scaled.py:
```diff
@@ -0,0 +1,283 @@
+    even_k_symbolic = (
+        sympy.gcd(sym_k, config.kwargs["BLOCK_K"]) == config.kwargs["BLOCK_K"]
+    )
+    allow_tf32 = torch.backends.cuda.matmul.allow_tf32 and (
```
I think this is for tf32 gemm, which is irrelevant here.

ipiszy commented on torch/_inductor/kernel/mm_scaled.py:
```diff
@@ -0,0 +1,283 @@
+
+    return dict(
+        GROUP_M=8,
+        EVEN_K=even_k_symbolic,
```
Could you help confirm how will this be used? Also, what's the alignment requirement for fp8 gemm from Triton and Cublas? Could you add some test cases for e.g. odd k, k divisible by 2 / 4 / 8 / 16? If it's not supported we should throw an error somewhere.

yangsiyu007 replied:
`EVEN_K` is used in the kernel template to load matrix blocks without masks when K is even. This is used in the mm's lowering as well.

Alignment requirement:

Eager
- Tensorwise Cublas: K, N divisible by 16 (will check if requirement on N is real, and if not will make a separate PR to update)
- Rowwise Cutlass: no requirement

Compiled (additional requirements from Triton which is auto-tuned w the eager/ATen choices) 
- Triton: no requirement on M

For the compiled case, requirement for K and N is checked in _meta_registrations already. I'll add a test case to test the error message. Added test cases for rowwise/tensorwise for when M is (1, 3, 33, 257, 1024), and K, N are 16 or a larger multiple of 32.

ipiszy replied:
"Tensorwise cublas": is the alignment requirement on N real?

"Rowwise Cutlass: no requirement": I'm surprised about this. I don't really believe that the CUTLASS fp8 gemm kernel doesn't have requirements on K. Are we sure about this?

yangsiyu007 replied:
Sorry I meant no additional requirements from Cutlass. Edited my comment above with more info (commenting out current 16-divisibility checks).

ipiszy commented on torch/_inductor/kernel/mm_scaled.py:
```diff
@@ -0,0 +1,283 @@
+        num_stages=config.num_stages,
+        num_warps=config.num_warps,
+        # tensor-wise scaling if scalar scales
+        SCALING_ROWWISE=len(scale_a.get_size()) != 0,
```
Should we also check that this is not blockwise? e.g. scale_a.size(0) == a.size(0).

ipiszy reviewed (commented):
Thanks @yangsiyu007 !

> Should the type of the accumulator ACC_TYPE always be in float32? If not, where is this type set (output layout?)?

I'd suggest to use fp32 as default. Currently no use cases for fp16 accumulation. 

> Should use_fast_accum be default False (as it is in aten/src/ATen/native/native_functions.yaml)? Probably should be False

As long as there is a way to set use_fast_accum it should be okay. For inference, most of cases we just use fast_accum.

yangsiyu007 committed:
Update UT for tensor-wise scaling. Get rid of allow_tf32

yangsiyu007 committed:
Add lowering UT for rowwise scaling

yangsiyu007 committed:
Skip K dim and BLOCK_M size combinations that cause Triton kernel to crash

yangsiyu007 committed:
Add input and scale dim checks and test them

vkuzo commented:
> Row-wise scaling allows operator fusion between preceding pointwise/reduction op and amax/cast:

thats great! asking as an inductor n00b and not for this PR, what would it take to support epilogue fusion as well for float8 gemms?

yangsiyu007 commented:
@vkuzo I should clarify that the fusion between the preceding op and the amax and cast is not considered "prologue fusion" for the scaled_mm op, which would mean that the preceding op is fused into the same kernel as scaled_mm (in P1477224245, there are still 2 kernels). AFAIK prologue fusion is not usually done for matmul ops, because it results in redundant computation of the preceding op each time a row/column is read by the matmul op. 

Epilogue fusion is already done by Inductor when one of the Triton kernels (configurations of the template) is selected by max-autotune. Here's an example of sigmoid + scaled_mm + relu: P1489769109. The relu is fused to the end of the scaled_mm kernel `tmp1 = triton_helpers.maximum(tmp0, acc)`, so there're still only 2 kernels in all. @ipiszy please correct me

yangsiyu007 committed:
Use more strict tolerance for eager vs compiled results comparisons

vkuzo commented on test/inductor/test_fp8.py:
```diff
@@ -346,6 +384,298 @@ def ln_fp8(x: Tensor, scale: Tensor, amax_buffer: Tensor):
+        )
+        self.assertEqual(y_eager.dtype, dtype)
+        self.assertEqual(y_compiled.dtype, dtype)
+        torch.testing.assert_close(y_eager, y_compiled, rtol=5e-2, atol=0)
```
the tolerance is higher than I would have expected.  If we compare, for example, the cuBLAS kernel (via torch._scaled_mm) versus the triton fp8 tutorial's matmul, is the tolerance similar?

yangsiyu007 replied:
I see one issue is that I include input quantization inside the compiled function, making the difference greater than if only the matmul is compiled. Updated the UTs to only include matmul in the compiled function, and all `rtol` to 0.01.

I also realized that I made the mistake of using `rand()` (Uniform[0, 1)) instead of `randn()` (Normal(0, 1)) to generate input matrices. Using `randn()`, some result elements are close to 0, so the relative difference can be quite large (for M, K, N = 2, 1024, 2048, one element is -0.0432 in the compiled case and -0.0234 via the cuBLAS kernel, 80% relative difference). So I've increased the `atol` to 0.05 or 0.07 to allow for this (see below on justification). 

> If we compare, for example, the cuBLAS kernel (via torch._scaled_mm) versus the triton fp8 tutorial's matmul, is the tolerance similar?

The Triton template used for the compiled case follows that of the mm kernel (with scaling and bias added), which is expressed slightly differently than the [tutorial's matmul](https://triton-lang.org/main/getting-started/tutorials/03-matrix-multiplication.html#sphx-glr-getting-started-tutorials-03-matrix-multiplication-py) but the approach is exactly the same. When I compared the cuBLAS kernel, tutorial's Triton and the templated Triton kernels, I did see larger relative difference (for element close to 0) between the templated Triton and cuBLAS for some shapes e.g. (3, 1024, 2048), (2, 1024, 2048. This turned out to be because certain configs (block M size etc) led to different results (expected as float addition is not associative - see my post in the internal Triton group), and once we use the same config for the tutorial's Triton kernel, then both Triton kernels had the same relative difference to cuBLAS. The difference is very small compared to how imprecise they are from the bfloat16 result. Repro script and output: P1496984171.

yangsiyu007 committed:
Move input quantization outside of compiled fn in UT

yangsiyu007 committed:
Use randn() in UT

yangsiyu007 committed:
Comment on atol (kernel config differences lead to slightly different result) and catch NoValidChoicesError

ipiszy reviewed (approved):
lgtm!

yangsiyu007 commented:
@pytorchbot merge

yangsiyu007 commented:
@pytorchbot merge

yangsiyu007 committed:
Merge branch 'main' into scaled-mm-lowering

yangsiyu007 commented:
@pytorchbot merge

yangsiyu007 committed:
Ignore untyped decorator register_lowering

yangsiyu007 committed:
Ignore typing for mm_scaled module

yangsiyu007 committed:
Ignore untyped decorator register_lowering correctly

Needed to sync with origin and run `with-proxy lintrunner -m origin/main` to reproduce.

yangsiyu007 commented:
@pytorchbot merge -f "Failing tests TestModuleCUDA.test_cpu_gpu_parity_nn_CTCLoss_cuda_float32 and TestModuleCPU.test_non_contiguous_tensors_nn_KLDivLoss_cpu_float32 are failing on main already."

ZainRizvi commented:
@pytorchbot revert -c ghfirst -m "Breaks internal tests. See D60568116"
